contracts/SDUtilityPool.sol

```diff
@@ -0,0 +1,906 @@
+     */
+    function requestWithdraw(uint256 _cTokenAmount) external override whenNotPaused returns (uint256 _requestId) {
+        if (_cTokenAmount > delegatorCTokenBalance[msg.sender] - delegatorWithdrawRequestedCTokenCount[msg.sender]) {
+            revert InvalidAmountOfWithdraw();
+        }
+        delegatorWithdrawRequestedCTokenCount[msg.sender] += _cTokenAmount;
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 sdRequested = (exchangeRate * _cTokenAmount) / DECIMAL;
+        _requestId = _requestWithdraw(sdRequested, _cTokenAmount);
+    }
+
+    /**
+     * @notice auxiliary method to put a withdrawal request, takes SD amount in input
+     * @param _sdAmount amount of SD to withdraw
+     * @return _requestId generated request ID for withdrawal
+     */
+    function requestWithdrawWithSDAmount(uint256 _sdAmount)
+        external
+        override
+        whenNotPaused
+        returns (uint256 _requestId)
+    {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 cTokenToReduce = (_sdAmount * DECIMAL) / exchangeRate;
+        if (cTokenToReduce > delegatorCTokenBalance[msg.sender] - delegatorWithdrawRequestedCTokenCount[msg.sender]) {
+            revert InvalidAmountOfWithdraw();
+        }
+        delegatorWithdrawRequestedCTokenCount[msg.sender] += cTokenToReduce;
+        _requestId = _requestWithdraw(_sdAmount, cTokenToReduce);
+    }
+
+    /**
+     * @notice finalize delegator's withdraw requests
+     */
+    function finalizeDelegatorWithdrawalRequest() external override whenNotPaused {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 maxRequestIdToFinalize = Math.min(nextRequestId, nextRequestIdToFinalize + finalizationBatchLimit) - 1;
+        uint256 requestId;
+        uint256 sdToReserveToFinalizeRequests;
+        for (requestId = nextRequestIdToFinalize; requestId <= maxRequestIdToFinalize; ) {
+            DelegatorWithdrawInfo memory delegatorWithdrawInfo = delegatorWithdrawRequests[requestId];
+            uint256 requiredSD = delegatorWithdrawInfo.sdExpected;
+            uint256 amountOfcToken = delegatorWithdrawInfo.amountOfCToken;
+            uint256 minSDRequiredToFinalizeRequest = Math.min(requiredSD, (amountOfcToken * exchangeRate) / DECIMAL);
+            if (
+                (sdToReserveToFinalizeRequests + minSDRequiredToFinalizeRequest > getPoolAvailableSDBalance()) ||
+                (delegatorWithdrawInfo.requestBlock + minBlockDelayToFinalizeRequest > block.number)
+            ) {
+                break;
+            }
+            delegatorWithdrawRequests[requestId].sdFinalized = minSDRequiredToFinalizeRequest;
+            sdRequestedForWithdraw -= requiredSD;
+            sdToReserveToFinalizeRequests += minSDRequiredToFinalizeRequest;
+            delegatorCTokenBalance[delegatorWithdrawInfo.owner] -= amountOfcToken;
```

Comment:
Would prefer if this is removed from the other mapping. 

Assume `delegatorCTokenBalance` is active cTokenBalance
and the other mapping to be `withdrawalInProgress` TokenBalance. 

Don't think we ever have to mix these two